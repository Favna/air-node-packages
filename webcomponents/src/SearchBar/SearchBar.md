### Usage of the SearchBar component:

**Examples:**

```tsx
import React from 'react'; 
import { SearchBar } from '@rws-air/webcomponents'; 

<SearchBar
  data-qa='table-search-bar'
  placeholder={ `Search...` }
  onChange={() => console.log('changed!')}
  onCancelSearch={() => console.log('Canceled Search!')}
/>
```

```jsx
import React from 'react'; 
import { SearchBar } from '@rws-air/webcomponents'; 

<SearchBar
  data-qa='table-search-bar'
  placeholder={ `Search...` }
  onChange={() => console.log('changed!')}
  onCancelSearch={() => console.log('Canceled Search!')}
/>
```

**Customizations**

```jsx
import React from 'react'; 
import { SearchBar } from '@rws-air/webcomponents'; 

<SearchBar
  data-qa='table-search-bar'
  placeholder={ `Search...` }
  onChange={() => console.log('changed!')}
  onCancelSearch={() => console.log('Canceled Search!')}
  paperElevation={5}
/>
```