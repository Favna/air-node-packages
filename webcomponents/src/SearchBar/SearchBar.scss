@import '../styles/vars';

.input {
  width: 100% !important;
}

.search {
  &Bar {
    height: 48px !important;
    display: flex !important;
    justify-content: space-between !important;
  }

  &IconButton {
    margin-right: -48px !important;
  }

  &Container {
    margin: auto 16px !important;
    width: calc(100% - 48px - 32px) !important;
  }
}

.icon {
  opacity: 0.54 !important;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) !important;

  &Button {
    opacity: 0.54 !important;
    transform: scale(1, 1) !important;
    transform: transform 200ms cubic-bezier(0.4, 0, 0.2, 1) !important;

    &Hidden {
      transform: scale(0, 0) !important;
      & > icon {
        opacity: 0 !important;
      }
    }

    &Disabled {
      opacity: 0.38 !important;
    }
  }
}
