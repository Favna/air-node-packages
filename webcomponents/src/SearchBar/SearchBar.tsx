import { grey } from '@material-ui/core/colors';
import IconButton from '@material-ui/core/IconButton';
import Input from '@material-ui/core/Input';
import Paper from '@material-ui/core/Paper';
import ClearIcon from '@material-ui/icons/Clear';
import SearchIcon from '@material-ui/icons/Search';
import classNames from 'classnames';
import React, { ChangeEvent, cloneElement, FC, KeyboardEvent, ReactElement, useState } from 'react';
import css from './SearchBar.scss';
import { dataQa } from 'typings';

export type SearchBarProps = {
  /** Whether to clear search on escape */
  cancelOnEscape?: boolean;
  /** Override or extend the styles applied to the component. */
  classes?: any;
  /** Custom top-level class */
  className?: string;
  /** Override the close icon. */
  closeIcon?: ReactElement;
  /** Override the search icon. */
  searchIcon?: ReactElement;
  /** Disables text field. */
  disabled?: boolean;
  /** Fired when the search is cancelled. */
  onCancelSearch?: () => any;
  /** Fired when the text value changes. */
  onChange?: (event: ChangeEvent<HTMLInputElement>) => any;
  /** Fired when the search icon is clicked. */
  onRequestSearch?: (event: any) => any;
  /** Fired when the search bar is focused */
  onFocus?: (event: any) => any;
  /** Fired when the focus on the search bar is lost */
  onBlur?: (event: any) => any;
  /** Fired when a key is released */
  onKeyUp?: (event: KeyboardEvent<HTMLInputElement>) => any;
  /** Sets placeholder text for the embedded text field. */
  placeholder?: string;
  /** Override the inline-styles of the root element. */
  style?: any;
  /** Override the paper element elevation */
  paperElevation?: number;
  /** data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
};

const SearchBar: FC<SearchBarProps> = (
  {
    className = '',
    disabled = false,
    placeholder = 'Search...',
    style = null,
    closeIcon = <ClearIcon style={{ color: grey[500] }} />,
    searchIcon = <SearchIcon style={{ color: grey[500] }} />,
    paperElevation = 2,
    ...props
  }: SearchBarProps
) => {
  const [value, setValue] = useState('');

  const handleFocus = (event: any): void => {
    if (props.onFocus) props.onFocus(event);
  };

  const handleBlur = (event: any): void => {
    if (value && value.trim().length === 0) setValue('');
    if (props.onBlur) props.onBlur(event);
  };

  const handleInput = (event: ChangeEvent<HTMLInputElement>): void => {
    setValue(event.target.value);
    if (props.onChange) props.onChange(event);
  };

  const handleCancel = (): void => {
    setValue('');
    if (props.onCancelSearch) props.onCancelSearch();
  };

  const handleRequestSearch = (): void => {
    if (props.onRequestSearch) props.onRequestSearch(value);
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (event.charCode === 13 || event.key === 'Enter') {
      handleRequestSearch();
    } else if (props.cancelOnEscape && (event.charCode === 27 || event.key === 'Escape')) {
      handleCancel();
    }

    if (props.onKeyUp) props.onKeyUp(event);
  };

  return (
    <Paper
      className={classNames(css.searchBar, className)}
      style={style}
      elevation={paperElevation}
    >
      <div className={css.searchContainer}>
        <Input
          onBlur={handleBlur}
          value={value}
          onChange={handleInput}
          onKeyUp={handleKeyUp}
          onFocus={handleFocus}
          fullWidth
          className={css.input}
          disabled={disabled}
          placeholder={placeholder}
          data-qa={props['data-qa']}
          disableUnderline
          {...props}
        />
      </div>
      <IconButton
        onClick={handleRequestSearch}
        classes={{
          root: classNames(css.iconButton, css.searchIconButton, {
            [css.iconButtonHidden]: value !== '',
          }),
          disabled: css.iconButtonDisabled,
        }}
        disabled={disabled}
      >
        {cloneElement(searchIcon, {
          classes: { root: css.icon },
        })}
      </IconButton>
      <IconButton
        onClick={handleCancel}
        classes={{
          root: classNames(css.iconButton, {
            [css.iconButtonHidden]: value === '',
          }),
          disabled: css.iconButtonDisabled,
        }}
        disabled={disabled}
      >
        {cloneElement(closeIcon, {
          classes: { root: css.icon },
        })}
      </IconButton>
    </Paper>
  );
};

export default SearchBar;