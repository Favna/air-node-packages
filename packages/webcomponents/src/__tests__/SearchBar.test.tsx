import AcUnitIcon from '@material-ui/icons/AcUnit';
import AddAPhotoIcon from '@material-ui/icons/AddAPhoto';
import { shallow } from 'enzyme';
import React from 'react';
import SearchBar from '../SearchBar/SearchBar';

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const searchBar = shallow(<SearchBar />);
    expect(searchBar).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const searchBar = shallow(
      <SearchBar
        cancelOnEscape closeIcon={<AcUnitIcon />} searchIcon={<AddAPhotoIcon />}
        paperElevation={3} disabled placeholder='Snaptest' data-qa='snap-qa'
      />
    );
    expect(searchBar).toMatchSnapshot();
  });
});