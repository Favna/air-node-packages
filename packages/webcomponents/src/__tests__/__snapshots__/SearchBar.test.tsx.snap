// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className=""
  elevation={3}
  style={Object {}}
>
  <div>
    <WithStyles(ForwardRef(Input))
      data-qa="snap-qa"
      disableUnderline={true}
      disabled={true}
      fullWidth={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      placeholder="Snaptest"
      value=""
    />
  </div>
  <WithStyles(ForwardRef(IconButton))
    classes={
      Object {
        "disabled": undefined,
        "root": "",
      }
    }
    disabled={true}
    onClick={[Function]}
  >
    <AddAPhotoIcon
      classes={
        Object {
          "root": undefined,
        }
      }
    />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(IconButton))
    classes={
      Object {
        "disabled": undefined,
        "root": "undefined",
      }
    }
    disabled={true}
    onClick={[Function]}
  >
    <AcUnitIcon
      classes={
        Object {
          "root": undefined,
        }
      }
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Paper))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className=""
  style={Object {}}
>
  <div>
    <WithStyles(ForwardRef(Input))
      disableUnderline={true}
      fullWidth={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      placeholder="Search..."
      value=""
    />
  </div>
  <WithStyles(ForwardRef(IconButton))
    classes={
      Object {
        "disabled": undefined,
        "root": "",
      }
    }
    onClick={[Function]}
  >
    <SearchIcon
      classes={
        Object {
          "root": undefined,
        }
      }
      style={
        Object {
          "color": "#9e9e9e",
        }
      }
    />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(IconButton))
    classes={
      Object {
        "disabled": undefined,
        "root": "undefined",
      }
    }
    onClick={[Function]}
  >
    <ClearIcon
      classes={
        Object {
          "root": undefined,
        }
      }
      style={
        Object {
          "color": "#9e9e9e",
        }
      }
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Paper))>
`;
